tests/keras/backend/backend_test.py

```diff
@@ -262,6 +262,53 @@ def ref_separable_conv(x, w1, w2, padding, data_format):
     return ref_conv(x2, w2, padding, data_format)
 
 
+def ref_pool(x, pool_size, strides, padding, data_format, pool_mode):
+    if data_format == 'channels_last':
+        if x.ndim == 3:
+            x = np.transpose(x, (0, 2, 1))
+        elif x.ndim == 4:
+            x = np.transpose(x, (0, 3, 1, 2))
+        else:
+            x = np.transpose(x, (0, 4, 1, 2, 3))
+
+    if padding == 'same':
+        pad = [(0, 0), (0, 0)] + [(s // 2, s // 2) for s in pool_size]
+        x = np.pad(x, pad, 'constant', constant_values=-np.inf)
+
+    y = []
+    for i in range(x.shape[0]):
+        _y = []
+        for j in range(x.shape[1]):
+            __y = []
+            for k in range(pool_size[0]):
+                k1 = None if pool_size[0] == (k + 1) else k + 1 - pool_size[0]
+                for l in range(pool_size[1]):
+                    l1 = None if pool_size[1] == (l + 1) else l + 1 - pool_size[1]
+                    if x.ndim == 5:
+                        for m in range(pool_size[2]):
+                            m1 = None if pool_size[2] == (m + 1) else m + 1 - pool_size[2]
+                            __y.append(x[i, j, k:k1:strides[0], l:l1:strides[1], m:m1:strides[2]])
+                    else:
+                        __y.append(x[i, j, k:k1:strides[0], l:l1:strides[1]])
+            __y = np.stack(__y, axis=-1)
+            if pool_mode == 'avg':
+                _y.append(np.mean(np.ma.masked_invalid(__y), axis=-1).data)
+            elif pool_mode == 'max':
+                _y.append(np.max(__y, axis=-1))
+        y.append(_y)
+    y = np.array(y)
```

Comment:
I wonder if this whole loop could be simplified somehow? It has a clear recursive structure. The same code should be generalizable to 1D/2D/3D/etc without having to manually unroll the loop.